modules/docs/src/05-design/02-schemas.md

```diff
@@ -0,0 +1,223 @@
+namespace example
+
+@trait 
+structure metadata {
+  @required
+  description: String
+}
+
+@metadata(description: "This is my own integer shape")
+integer MyInt 
+```
+
+When processing this smithy model, smithy4s renders a `case class Metadata(description: String)`, with an assocaited `ShapeTag[Metadata]` instance, and the following expression in the companion object of `MyInt` : 
+
+```scala
+val hints = Hints(
+  Metadata("this is my own integer shape)
+)
+```
+
+The `smithy4s.Hints` type is a polymorphic map that can hold shapes, keyed by `ShapeTag`. A `ShapeTag` is basically a uniquely identified tag that uses referential equality. Every schema can hold a `Hints` instance, which means that in additiona to the datatype structures, Schemas also offer an accurate reflection of the trait values that annotate shapes in the smithy models. 
+
+Smithy4s uses these hints to implement interpreters. For instance, the `smithy.api#jsonName` smithy trait translates to a `smithy.api.JsonName` Scala type, that we can query from a `Hints` instance when implementing a `Schema ~> JsonCodec` transformation. This allows to give users a little customisability in 
+the json serialisation of their datatypes. 
+
+
+
+#### Structures  
+
+A structure, also referred to as product, or record, is a construct that groups several values together. Typically, it translates naturally to a case class. 
+
+```kotlin
+namespace example 
+
+structure Foo {
+  @required 
+  a: Integer,
+  @length(min: 1)
+  b: String, 
+}
+```
+
+... and the associated, generated Scala code: 
+
+```scala
+package example
+
+import smithy4s.schema.Schema._
+
+case class Foo(a: Int, b: Option[String] = None)
+
+object Foo extends smithy4s.ShapeTag.Companion[Foo] {
+  val id: smithy4s.ShapeId = smithy4s.ShapeId("example", "Foo")
+
+  implicit val schema: smithy4s.Schema[Foo] = struct(
+    int.required[Foo]("a", _.a),
+    string.optional[Foo]("b", _.b).addHints(smithy.api.Length(Some(1), None)) 
+  ){
+    Foo.apply
+  }.withId(id)
+}
+```
+
+As you can see, the smithy **structure** translates quite naturally to a Scala **case-class**. 
+Every member of the structure that does not have the `@required` trait is rendered as an Optional value defaulting to None (by default, smithy4s sorts the fields before rendering the case class so that the required ones appear before the optional ones. That is a pragmatic decision that tends to improve UX for users.)
+
+Indeed, for each field, there is an associated reference to a schema (int, string, ...), a string label, and a lambda calling the case class accessor that allows the retrieval of the associated field value. Additionally, the constructor of the case class is also referenced in the Schema. 
+
+Typically, the accessors are needed for encoding the data, which involves destructuring it to access its individual components. The labels are there to cater to serialisation mechanisms like Json or Xml, which apply some labelled nesting on component data. 
```

Comment:
> which apply some labelled nesting on component data

I'm not quite clear what is meant by this

Reply:
I've rephrased, let me know how it reads